service/task_manager_module.cc

```diff
@@ -50,6 +51,8 @@ static bool tablet_id_provided(const locator::tablet_task_type& task_type) {
 }
 
 future<std::optional<tasks::virtual_task_hint>> tablet_virtual_task::contains(tasks::task_id task_id) const {
+    co_await _ss._migration_manager.local().get_group0_barrier().trigger();
```

Comment:
@kbr-scylla can you please confirm/deny that this solves the problem described in the patch notes?

Reply:
Hard to say without understanding the context.

What are the causality dependencies here? From the description I'm guessing that:
- **something** guarantees that the `task_id` we're looking for is already in `system.tablets` (what is that something?) in the context of this PR
- but that same thing does not guarantee that it's already in the in-memory data structures that are built from `system.tablets` when we reload the table (so the table was not reloaded)

if first point holds, then read barrier will make sure that whatever was visible in system.tablets is also visible in the in-memory state, but note that we only are guaranteed to see **at least as recent** state, i.e. the state we see could be e.g. after the data was removed (we're basically sleeping for a potentially long time) and then we won't see the data, and then I don't know if it's guaranteed that the data is still there in the context we're discussing in, so dunno.

Reply:
> something guarantees that the task_id we're looking for is already in system.tablets (what is that something?) in the context of this PR

This something is an API request for starting a new repair. This will create an entry in group0, which gets to `system.tablets`, but not immediately visible in memory, because the topology object is regenerated first.

Reply:
The scenario which prompted this PR is:
* user starts repair via API
* immediately after previous request returned, user checks state of repair task

Before this PR, user would get an error that repair task with given ID doesn't exist. Upon retry, they would see the expected task.

Reply:
> This will create an entry in group0, which gets to system.tablets, but not immediately visible in memory

How does it create the entry? If it uses `raft_group0_client::add_entry` API to commit the entry, this underneath calls `raft::server::add_entry` with `wait_type::applied`, meaning that the future will resolve only after the entry is **completely applied locally**, which includes any in-memory changes that application of the entry performs.

Reply:
> If it uses raft_group0_client::add_entry API to commit the entry, this underneath calls raft::server::add_entry with wait_type::applied, meaning that the future will resolve only after the entry is completely applied locally, which includes any in-memory changes that application of the entry performs.

So if the repair API calls `add_entry` and waits for the future before returning to the user, then the bug must be somewhere else

Reply:
Unless the problem is that the first entry added to group 0 doesn't actually update system.tablets, but some kind of background process starts and a follow-up entry is then added (maybe on topology coordinator) but then this PR also won't solve the problem as the read_barrier might observe the state in-between those 2 entries.

Reply:
> How does it create the entry? If it uses raft_group0_client::add_entry API to commit the entry, this underneath calls raft::server::add_entry with wait_type::applied, meaning that the future will resolve only after the entry is completely applied locally, which includes any in-memory changes that application of the entry performs.

It starts the repair task and gets the id directly from system.tablets

Reply:
> starts the repair task

ok but does the API to start repair task, which is used by the test, wait for the group 0 entry to get applied (does it wait for the future returned by `add_entry` to get resolved)? And then does the test wait for the API to return response? If yes, then by that moment both system.tables and in-memory data structures should already be updated.

Reply:
Tablet repair API waits for `add_entry` but returns after the repair is finished. So the test does not wait for it (or until the entry is applied), but rather until `system.tablets` contains the repair info.